java/src/jmri/jmrix/bidib/BiDiBTrafficController.java

```diff
@@ -0,0 +1,1487 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import jmri.CommandStation;
+import jmri.NmraPacket;
+import jmri.jmrix.PortAdapter;
+
+import org.bidib.jbidibc.messages.BidibLibrary;
+import org.bidib.jbidibc.messages.exception.ProtocolException;
+import org.bidib.jbidibc.messages.utils.ByteUtils;
+
+import org.bidib.jbidibc.core.BidibMessageProcessor;
+import org.bidib.jbidibc.core.BidibInterface;
+import org.bidib.jbidibc.messages.BidibPort;
+import org.bidib.jbidibc.messages.ConnectionListener;
+import org.bidib.jbidibc.core.DefaultMessageListener;
+import org.bidib.jbidibc.messages.Feature;
+import org.bidib.jbidibc.messages.LcConfig;
+import org.bidib.jbidibc.messages.LcConfigX;
+import org.bidib.jbidibc.core.MessageListener;
+import org.bidib.jbidibc.messages.base.RawMessageListener;
+import org.bidib.jbidibc.messages.Node;
+import org.bidib.jbidibc.core.NodeListener;
+import org.bidib.jbidibc.messages.ProtocolVersion;
+import org.bidib.jbidibc.messages.StringData;
+import org.bidib.jbidibc.messages.helpers.Context;
+import org.bidib.jbidibc.core.node.listener.TransferListener;
+import org.bidib.jbidibc.messages.exception.PortNotFoundException;
+import org.bidib.jbidibc.core.node.BidibNode;
+import org.bidib.jbidibc.messages.message.BidibCommandMessage;
+import org.bidib.jbidibc.core.node.BidibNodeAccessor;
+import org.bidib.jbidibc.messages.utils.NodeUtils;
+import org.bidib.jbidibc.messages.enums.CommandStationState;
+import org.bidib.jbidibc.messages.enums.LcOutputType;
+import org.bidib.jbidibc.messages.enums.PortModelEnum;
+import org.bidib.jbidibc.messages.message.AccessoryGetMessage;
+import org.bidib.jbidibc.messages.message.BidibRequestFactory;
+import org.bidib.jbidibc.messages.message.CommandStationSetStateMessage;
+import org.bidib.jbidibc.messages.message.FeedbackGetRangeMessage;
+import org.bidib.jbidibc.core.node.CommandStationNode;
+import org.bidib.jbidibc.core.node.BoosterNode;
+import org.bidib.jbidibc.messages.BoosterStateData;
+import org.bidib.jbidibc.messages.enums.BoosterControl;
+import org.bidib.jbidibc.messages.enums.BoosterState;
+import org.bidib.jbidibc.messages.enums.CommandStationProgState;
+import org.bidib.jbidibc.messages.port.BytePortConfigValue;
+import org.bidib.jbidibc.messages.port.PortConfigValue;
+import org.bidib.jbidibc.messages.port.ReconfigPortConfigValue;
+import org.bidib.jbidibc.simulation.comm.SimulationBidib;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The BiDiB Traffic Controller provides the interface for JMRI to the BiDiB Library (jbidibc) - it
+ * does not handle any protocol functions itself. Therefor it does not extend AbstractMRTrafficController.
+ * Instead, it delegates BiDiB handling to a BiDiB controller instance (serial, simulation, etc.) using BiDiBInterface.
+ * 
+ * @author Bob Jacobsen Copyright (C) 2002
+ * @author Eckart Meyer Copyright (C) 2019-2023
+ *
+ */
+@SuppressFBWarnings(value = "JLM_JSR166_UTILCONCURRENT_MONITORENTER")
```

Comment:
What does this `@SuppressFBWarnings` actually suppress?

Quote from https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html
> JLM: Synchronization performed on util.concurrent instance (JLM_JSR166_UTILCONCURRENT_MONITORENTER)
> 
> This method performs synchronization on an object that is an instance of a class from the java.util.concurrent package (or its subclasses). Instances of these classes have their own concurrency control mechanisms that are orthogonal to the synchronization provided by the Java keyword synchronized. For example, synchronizing on an AtomicBoolean will not prevent other threads from modifying the AtomicBoolean.
> 
> Such code may be correct, but should be carefully reviewed and documented, and may confuse people who have to maintain the code at a later date.

Reply:
I've added a comment that discusses how these AtomicBooleans are being used for synchronization.